go/obscuronode/enclave/sql/edgelessdb.go

```diff
@@ -0,0 +1,480 @@
+	"github.com/go-sql-driver/mysql"
+)
+
+/*
+   Getting an edgeless DB connection is a bit of a dance in order to achieve mutual trust.
+
+   The edgeless DB is unusable when it's first started, as an owner, we must initially:
+     - do a remote attestation on the report provided by {edbAddress}/quote
+     - create a ca cert that will authenticate our DB users going forwards
+     - prepare a manifest.json that contains that CA cert and some SQL to initialise the DB tables and user
+     - submit that manifest.json file to {edbAddress}/manifest, using the certificate provided from /quote to authenticate
+     - seal and persist the manifest.json and the certs so we can reconnect if enclave is restarted
+
+   When connecting to an already-initialized edgeless DB we must:
+     - perform remote attestation on the edgeless db
+     - unseal the manifest.json and get the hash of it, also unseal the certificate that edb was initialized with
+	 - verify the /signature request for edgeless DB matches the manifest.json hash
+     - connect to edb with the persisted cert - it's now safe to read and write to the DB
+
+	Some useful documentation for this:
+		Main edb docs: https://docs.edgeless.systems/edgelessdb/#/
+		EDB demo docs: https://github.com/edgelesssys/edgelessdb/tree/main/demo
+		// Note: due to an issue with the dependency, I've duplicated the relevant parts of the ERA tool into edbattestation.go
+		ERA - remote attestation tool: https://github.com/edgelesssys/era
+*/
+
+var manifestSQLStatements = []string{
+	"CREATE USER obscuro REQUIRE ISSUER '/CN=obscuroCA' SUBJECT '/CN=obscuroUser'",
+	"CREATE DATABASE obsdb",
+	"CREATE TABLE obsdb.keyvalue (ky varbinary(64) primary key, val blob)",
+	"GRANT ALL ON obsdb.keyvalue TO obscuro",
+}
+
+const (
+	dataDir                = "/data/"
+	manifestFilepath       = dataDir + "manifest.json"
+	caCertFilepath         = dataDir + "ca-cert.pem"
+	userCertFilepath       = dataDir + "user-cert.pem"
+	userKeyFilepath        = dataDir + "user-key.pem"
+	attestationCfgFilepath = dataDir + "edgelessdb-sgx.json"
+	edbHTTPPort            = "8080"
+)
+
+type manifest struct {
+	SQL   []string `json:"sql"`
+	Cert  string   `json:"ca"`
+	Debug bool     `json:"debug"`
+}
+
+type EdgelessDBConfig struct {
+	Host string
+}
+
+func EdgelessDBConnector(edbCfg EdgelessDBConfig) (ethdb.Database, error) {
+	// before we try to connect to the Edgeless DB we have to do remote attestation on it
+	edbPEM, err := performRemoteAttestation(edbCfg.Host)
+	if err != nil {
+		return nil, fmt.Errorf("remote attestation of edgeless DB failed - %w", err)
+	}
+	log.Info("retrieved edb PEM: %s", edbPEM)
+
+	// IF DEBUGGING: it can be useful to persist the edb.pem so you can connect to edb from mysql-client on the container
+	//_ = sealAndPersist(edbPEM, dataDir+"edb.pem")
+
+	// now we know we are talking to a secure enclave, we can get the manifest and connect (or initialise if first time)
+	manifest, err := readManifestIfExists()
```

Comment:
Just wondering here. If there is a manifest already stored, it means that the OE has already performed the RA.
So in theory the RA step would not be necessary, right? 
Is there something in their docs covering this?

Reply:
Also wondering, what happens if after the restart, the OE performs the RA and the edb returns a different cert. We'd basically have one cert in the stored manifest, and another returned.

Reply:
My understanding is we'd need to remote attest before connecting to it every time. It might have been hacked since you last connected, you need to guarantee it's running the certified EdgelessDB code rather than just happens to have access to the certificates before you start spilling your secrets.

The cert that's in the manifest is *our* CA that we generated so that should be static for the lifetime of our enclave. We don't persist anything related to their EDB cert, if it changes we should be able to keep chatting to them securely (confident from the attestation and the manifest hash that they can be trusted).

Reply:
the only case where it might fail is where intel would say that this particular hardware is no longer secure. But that would also make the OE hardware insecure.